Draft new ADR on numeric LPA references (#160)
ministryofjustice/opg-data-lpa-uid: docs/architecture/decisions/0006-unique-ids-for-private-beta.md

# 1. Unique IDs for private beta

Date: 2024-10-08

Edited: 2024-10-08

## Status

Proposed

## Context

Originally (see [ADR 0004](0004-generating-unique-ids.md)) we looked at randomly generated alpha-numeric reference numbers for the modernised LPAs. The original constraints at the time were:

The reference numbers generated must:

- Only use alpha-numeric characters for readability purposes
- Be unique to allow us to identify 1 LPA application per reference number

> - Have a clear prefix to differentiate them from existing IDs and other codes used in out processes.

Reviewer: ```suggestion
- Have a clear prefix to differentiate them from existing IDs and other codes used in our processes.
```

> - Reduce the risk on iteration attacks by making them non-sequential and less guessable (if a trheat actor knows LPA 0001 exists trying 0002 gets a result)

Reviewer: ```suggestion
- Reduce the risk on iteration attacks by making them non-sequential and less guessable (if a threat actor knows LPA 0001 exists trying 0002 gets a result)
```

LPA reference numbers will be reduced to the numeric components in line with business requirements.

Thus a valid LPA number would be M-0000-0000-0013 where the 3 is a checkdigit generated by the Damm algorithm.

Typical tranposing issues such as digit order twiddling. i.e. 0000-0000-1234 is a correct checkdigit of 4, but 0000-0000-2134 would be invalid detecting the 12 to 21 swap.

We have estimated scenarios for potential ID usage over a 10 year period at current and aggressive growth. (see [internal spreadsheet](https://docs.google.com/spreadsheets/d/1kH1BOk8iOTUg3aEphD5J18LV4zhr4mQn/edit?usp=sharing&ouid=104236485660673640558&rtpof=true&sd=true)).

## Consequences

- Reducing the information space around the reference number to only the 0-9 characters with a check digit means we have a maximum of 99,999,999,999 numbers available.

- We will generate references non-sequentially, so that it is easier to detect enumeration attacks on any service that uses the ref number as an ID. i.e. 0001 is not followeed by 0002 and our approach should inject sufficient padding to make iteration attacks easily detectable.

> - LPA reference numbers that are purely numeric are more guessable with enough tries or given sufficinet requests. They should not be used as publically facing identifiers in web based services without some 2nd factor (for example the paper verification codes) and any services using them as url identifiers or form parameters should put in place detection for brute forcing and iteration based attacks.

Reviewer 2: ```suggestion
- LPA reference numbers that are purely numeric are more guessable with enough tries or given sufficient requests. They should not be used as publicly facing identifiers in web based services without some 2nd factor (for example the paper verification codes) and any services using them as url identifiers or form parameters should put in place detection for brute forcing and iteration based attacks.
```

- LPA Reference nymbers should not be used in URLs on external-facing systems. For external-facing APIs or services (such as the 3rd party supporter API we plan) that need to make direct reference to an LPA, a local UUID associated with an LPA should be used instead.

- During private beta it is essential we track LPA reference number usage and adjust our approach accordingly. Particularly if more drafts are created than expected as this could use up the available set of IDs at a greater rate. A checkpoint at the end of private beta to assess this should be put in place.

- The reduced information space will mean we may have to use a different technology in the UID generator due to likelihood of collisions from random generation, potentially a traditional RDS with a sequence generator. This will create rework for teams.

- If ID space looks like it is close to exhausted, OPG has the option to implement a successor prefix i.e. M-0000-0000-0001 to L-0000-0000-0001.

> - The prefix is part of the LPA reference number and should asked for and stored in all interactions cases. The reference number should not be stored in a database without the prefix, so should be stored as a string. Dashes are considered part of the string for the performance benefit not constantly re-formatting provides. An LPA reference number is therefore: "M-0000-0000-1234" not "000000001234"

Reviewer 3: Is the prefix a part of the checksum calculation?

Author: Damm only handles numerics, so only the numeric parts form the checksum.

We have estimated scenarios for potential ID usage over a 10 year period at current and aggressive growth. (see [spreadsheet attached to issue MLPA-2450](https://opgtransform.atlassian.net/browse/MLPAB-2450)).

## Consequences

- Reducing the information space around the reference number to only the 0-9 characters with a check digit means we have a maximum of 99,999,999,999 numbers available.

- We will generate references non-sequentially, so that it is easier to detect enumeration attacks on any service that uses the ref number as an ID. i.e. 0001 is not followeed by 0002 and our approach should inject sufficient padding to make iteration attacks easily detectable and avoid predictable step jumps.

- LPA reference numbers that are purely numeric are more guessable with enough tries or given sufficient requests. They should not be used as publicly facing identifiers in web based services without some 2nd factor (for example the paper verification codes) and any services using them as url identifiers or form parameters should put in place detection for brute forcing and iteration based attacks.

> - They should not be used in URLs on external-facing systems. In external facing services or APIs that need to make direct reference to an LPA the UUID associated with an LPA should be used instead.

Reviewer 2: LPAs don't currently have a UUID. There's a UUID created in MRLPA before the LPA is assigned a reference number, but it's not shared into any other systems.

At the moment, are "external facing services or APIs" only MRLPA and Use?

Author: In terms of External-facing I was thinking more the MrLPA API that will go live for supporting orgs at some point, rather than internal OPG systems like MrLPA and Use.

- During private beta it is essential we track LPA reference number usage and adjust our approach accordingly. Particularly if more drafts are created than expected as this could use up the available set of IDs at a greater rate. A checkpoint at the end of private beta to assess this should be put in place.

- The reduced information space will mean we may have to use a different technology in the UID generator due to likelihood of collisions from random generation, potentially a traditional RDS with a sequence generator. This will create rework for teams.

> - If ID space looks like it is close to exhausted, OPG has the option to implement a successor prefix i.e. M-0000 0000 0001 to L-0000 0000 0001.

Reviewer 2: ```suggestion
- If ID space looks like it is close to exhausted, OPG has the option to implement a successor prefix i.e. M-0000-0000-0001 to L-0000-0000-0001.
```

> - The prefix is part of the LPA reference number and should asked for and stored in all interactions cases. The reference number should not be stored in a database without the prefix, so should be stored as a string. Formatting, such as dashes or spaces to improve readability are not part of the data, so are not stored.

Reviewer 2: We currently do store dashes as part of the reference number and, from experience with Sirius, it's a lot easier to do so because otherwise you're constantly reformatting it. e.g. Sirius stores legacy reference numbers like 700212345678 but converts them into 7002-1234-5678 on every database fetch because that's the only way they're ever displayed or used.

Author: Given we're already a way down that path I'll remove that note then.

- We will generate references non-sequentially, so that it is easier to detect enumeration attacks on any service that uses the ref number as an ID. i.e. 0001 is not followeed by 0002 and our approach should inject sufficient padding to make iteration attacks easily detectable and avoid predictable step jumps.

- LPA reference numbers that are purely numeric are more guessable with enough tries or given sufficient requests. They should not be used as publicly facing identifiers in web based services without some 2nd factor (for example the paper verification codes) and any services using them should put in place detection for brute forcing and iteration based attacks.

> - They should not be used in URLs on external-facing systems. None OPG Digital facing services or external-facing APIs (such as the 3rd party supporter API we plan) that need to make direct reference to an LPA a local UUID associated with an LPA should be used instead.

Reviewer 2: I think this is clearer, if I've carried the meaning?
```suggestion
- They should not be used in URLs on external-facing systems. For OPG Digital facing services or external-facing APIs (such as the 3rd party supporter API we plan) that need to make direct reference to an LPA, a local UUID associated with an LPA should be used instead.
```
